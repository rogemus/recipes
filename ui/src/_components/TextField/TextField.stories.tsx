import { fn } from "@storybook/test";

import TextField from "./TextField";

import type { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof TextField> = {
  component: TextField,
  title: "Form/TextField",
  args: {
    id: "id",
    name: "Email",
    type: "email",
    label: "Email",
    placeholder: "Email ...",
    onBlur: fn(),
    onChange: fn(),
  },
};

export default meta;

type Story = StoryObj<typeof TextField>;

export const Default: Story = {};
