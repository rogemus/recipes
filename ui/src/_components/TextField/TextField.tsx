import { type FC } from "react";

import FieldError from "../FieldError";

import styles from "./TextField.module.css";

import type { TextFieldProps } from "./TextField.types";

const TextField: FC<TextFieldProps> = ({
  defaultValue,
  error,
  id,
  label,
  name = "",
  placeholder,
  testId = "Input",
  type = "text",
  required,
  ref,
  onChange,
  onBlur,
}) => {
  return (
    <div className={styles.wrapper} data-testid={testId}>
      <label
        className={styles.label}
        data-testid={`${testId}-label`}
        htmlFor={id}
      >
        {label}
      </label>
      <input
        data-testid={`${testId}-input`}
        className={styles.input}
        type={type}
        id={id}
        name={name}
        {...(ref && { ref })}
        {...(onChange && { onChange })}
        {...(onBlur && { onBlur })}
        {...(required && { required: true })}
        {...(placeholder && { placeholder })}
        {...(defaultValue && { defaultValue })}
      />
      <FieldError error={error} testId={`${testId}-error`} />
    </div>
  );
};

export default TextField;
