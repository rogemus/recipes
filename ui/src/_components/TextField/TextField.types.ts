import type { FieldValues } from "react-hook-form";
import type {
  ChangeHandler,
  RefCallBack,
  UseFormRegister,
} from "react-hook-form";
import type { Comp } from "@/_models";

export type TextFieldProps = {
  defaultValue?: string;
  disabled?: boolean;
  error?: string;
  id: string;
  label: string;
  name: string;
  placeholder?: HTMLInputElement["placeholder"];
  required?: boolean;
  type?: HTMLInputElement["type"];
  register?: UseFormRegister<FieldValues>;
  onChange?: ChangeHandler;
  onBlur?: ChangeHandler;
  ref?: RefCallBack;
} & Comp;
