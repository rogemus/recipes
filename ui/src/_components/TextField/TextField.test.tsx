import { expect, test, describe, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import TextField from "./TextField";

import type { TextFieldProps } from "./TextField.types";

const defaultProps: TextFieldProps = {
  label: "test-label",
  id: "test-id",
  name: "test-name",
  placeholder: "Test placeholder",
  error: "Test error",
};

describe("@component/TextField", () => {
  test("renders component correctly", () => {
    const { container } = render(<TextField {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  test("onChange triggered", async () => {
    const mockOnChange = vi.fn();
    const user = userEvent.setup();

    render(<TextField {...defaultProps} onChange={mockOnChange} />);
    const input = screen.getByTestId("Input-input");

    await user.type(input, "Test Content");

    expect(mockOnChange).toHaveBeenCalled();
    expect(input).toHaveValue("Test Content");
  });

  test("onBlur triggered", async () => {
    const mockOnBlur = vi.fn();

    render(<TextField {...defaultProps} onBlur={mockOnBlur} />);
    const input = screen.getByTestId("Input-input");

    input.focus();
    expect(input).toHaveFocus();

    input.blur();
    expect(mockOnBlur).toHaveBeenCalledOnce();
  });
});
