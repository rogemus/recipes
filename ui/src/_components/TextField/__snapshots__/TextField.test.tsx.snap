// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@component/TextField > renders component correctly 1`] = `
<div>
  <div
    class="_wrapper_ec2ce5"
    data-testid="Input"
  >
    <label
      class="_label_ec2ce5"
      data-testid="Input-label"
      for="test-id"
    >
      test-label
    </label>
    <input
      class="_input_ec2ce5"
      data-testid="Input-input"
      id="test-id"
      name="test-name"
      placeholder="Test placeholder"
      type="text"
    />
    <div
      data-testid="Input-error"
    >
      <p>
        Test error
      </p>
    </div>
  </div>
</div>
`;
